src/anml/data/data.py

```diff
@@ -168,21 +176,51 @@ def process_params(self, df: pd.DataFrame):
         
         design_mat_blocks = []
         constr_mat_blocks = []
+        re_mat_groups = defaultdict(dict)
         lbs = []
         ubs = []
+        self.fe_variables_names = []
         for param_set in self._param_set:
             for parameter in param_set.parameters:
                 for variable in parameter.variables:
-                    design_mat_blocks.append(variable.design_mat(df=df))
+                    design_mat = variable.design_mat(df=df)
+                    design_mat_blocks.append(design_mat)
+                    var_name = parameter.param_name + '_' + variable.covariate
+                    self.fe_variables_names.append(var_name)
+                    if variable.add_re:
+                        re_mat_groups[variable.col_group][var_name] = design_mat
                     mat, lb, ub = variable.get_constraint_matrix()
                     constr_mat_blocks.append(mat)
                     lbs.append(lb)
                     ubs.append(ub)
-        design_matrix = np.hstack(design_mat_blocks)
-        constr_matrix = block_diag(*constr_mat_blocks)
-        lower_bounds = np.hstack(lbs)
-        upper_bounds = np.hstack(ubs)
-        assert design_matrix.shape[1] == constr_matrix.shape[1]
-        assert len(lower_bounds) == len(upper_bounds) == constr_matrix.shape[0]
-
-        return design_matrix, constr_matrix, lower_bounds, upper_bounds
+        self.design_matrix = np.hstack(design_mat_blocks)
+        self.constr_matrix = block_diag(*constr_mat_blocks)
+        self.constr_lower_bounds = np.hstack(lbs)
+        self.constr_upper_bounds = np.hstack(ubs)
+        assert self.design_matrix.shape[1] == self.constr_matrix.shape[1]
+        assert len(self.constr_lower_bounds) == len(self.constr_upper_bounds) == self.constr_matrix.shape[0]
+
+        if len(re_mat_groups) == 0:
+            self.re_matrix = None
+        else:
+            self.groups_info = defaultdict(dict)
+            re_mat_blocks = []
+            self.re_variables_names = []
+            for col_group, dct in re_mat_groups.items():
+                self.encode_groups(col_group, df)
+                grp_assign = [self.groups_info[col_group][g] for g in df[col_group]]
+                n_group = len(self.groups_info[col_group])
+
+                self.re_variables_names.append(dct.keys())
+                mat = np.hstack(dct.values())
+                n_coefs = mat.shape[1]
+                re_mat = np.zeros((mat.shape[0], n_coefs * n_group))
+                for i, row in enumerate(mat):
+                    grp = grp_assign[i]
+                    re_mat[i, grp * n_coefs: (grp + 1) * n_coefs] = row 
+                re_mat_blocks.append(re_mat)
+            
+            self.re_matrix = block_diag(*re_mat_blocks)
```

Comment:
Here I am a little confused, organize random effect design matrix as a block diagonal matrix by different `col_group`?

Reply:
yes, it's a bit easier for me to organize random effects variables that way (in terms of writing the code), having all random effects from one grouping together and then stack with those from other groupings. I can explain to you in words when we talk (which would be much easier than typing)

Reply:
maybe I can try to explain a bit. I was thinking something like `y = X*beta + Z_1u_1 + Z_2u2` where `Z_1` is the random effects matrix corresponding to one grouping (or `col_group`) and `Z_1` itself can also be block according to groups, and likewise `Z_2` is for another grouping. Then to combine them I stack the `u`s and use block diagonal for the `Z`s. (if this still does not make sense let's talk!)

Reply:
`Z_1` and `Z_2` are block diagonal make sense, and they should have the same number of rows (corresponding to the number of observations). But after stacking `Z_1` and `Z_2` as a bigger block diagonal matrix, how can we use this matrix? I imagine multiple a whole vector of random effects and reshape the random effects. But I also afraid it will be really big.

Reply:
what do you mean by reshaping? so I guess in the general case, if we have two groupings that are independent of each other, and depending on which grouping we choose to sort the rows, `Z_i` may be block diagonal or maybe not. when we solve the model it would just be some `Z` matrix, whatever the structure is. it indeed maybe large (though from the forecast team thing, numpy seems to handle that okay.)

Reply:
for example if `Z = block_diag(Z_1, Z_2)` and `u = np.hstack((u_1, u_2))` and after we do `v = Z.dot(u)`, `v` will have size `2*num_obs`, and we need to reshape it for it to do calculation with the residual `r = y - X.dot(beta) - np.sum(v.reshape(2, num_obs), axis=0)`.
Maybe I understand it not correctly. The alternative is to make `Z` a list of Numpy array.

Reply:
oooh, I see. I meant a `hstack`, like `[Z_1, Z_2]`. sorry! does that make more sense now?

Reply:
yes!!!